### PR TITLE
SQL: Allow partial results in SQL queries

### PR DESCRIPTION
**Pull request by bpintea**
This adds support for partial results to SQL.
The lenient mode is controlled by a new query parameter,
`allow_partial_search_results`, false by default. On shard failures, the
errors are added as Warning headers to the response. Only a first set of
failures are sent to the client, the last header briefs on the number of
remaining suppressed ones.

Fixes #33148.

Preview: https://elasticsearch_85897.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/sql-jdbc.html

### CODE REVIEW

File: docs/reference/sql/apis/sql-search-api.asciidoc [hunk 1]
```diff
@@ -54,6 +54,11 @@ precedence.
 [[sql-search-api-request-body]]
 ==== {api-request-body-title}
 
+`allow_partial_search_results`::
+(Optional, Boolean) If `true`, returns partial results if there are shard
+request timeouts or shard failures. If `false`, returns an error with no
```
Comment (luigidellaquila) [inline docs/reference/sql/apis/sql-search-api.asciidoc:59]: I see this comment is slightly different in https://www.elastic.co/guide/en/elasticsearch/reference/current/search-search.html#search-search-api-query-params, ie. it has a link with `<<shard-failures,shard failures>>`.
Would it work here? Or maybe it's in a different module and it doesn't resolve the link?

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/plan/logical/command/sys/SysColumnsTests.java [hunk 2]
```diff
@@ -287,7 +287,8 @@ private int executeCommandInOdbcModeAndCountRows(String sql) {
             false,
             false,
             null,
-            null
+            null,
+            false
```
Comment (luigidellaquila) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/plan/logical/command/sys/SysColumnsTests.java:291]: Maybe these can be randomized (not that it makes much sense to pass `true` for these commands, just to validate that the input is accepted correctly)

**Review by luigidellaquila — APPROVED** [review]
LGTM, just left a couple of comments

**Review by costin — APPROVED** [review]
Left some cosmetic comments, LGTM otherwise.

**Review by astefan — APPROVED** [review]
LGTM
Left some minor comments/questions.
Also, have you considered the CLI in any way for this new parameter, or it doesn't make sense?

File: docs/reference/sql/apis/sql-search-api.asciidoc (hunk 1 above)
Comment (bpintea) [reply docs/reference/sql/apis/sql-search-api.asciidoc]: Link to shard failures now included.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/plan/logical/command/sys/SysColumnsTests.java (hunk 2 above)
Comment (bpintea) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/plan/logical/command/sys/SysColumnsTests.java:291]: I guess these values were inherited from `SqlTestUtils.TEST_CFG`; some of these could be indeed randomised, but to stay consistent more changes would be needed in this class. I would limit this PR's scope.

File: docs/reference/sql/endpoints/jdbc.asciidoc [hunk 3]
```diff
@@ -155,6 +155,12 @@ queries execute on the data in the local cluster only.
 +
 experimental:[] See <<modules-cross-cluster-search,{ccs}>>.
 
+[discrete]
+==== Error handling
+
+`allow.partial.search.results` (default `false`):: How to handle shard failures: return partial results, if `true`, or fail the
```
[the remark this answers is not in the document]
Comment (bpintea) [reply docs/reference/sql/endpoints/jdbc.asciidoc]: Suggestion adopted with slight change.

File: x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/JdbcShardFailureIT.java [hunk 4]
```diff
@@ -60,10 +62,79 @@ public void createTestIndex() throws IOException {
         client().performRequest(request);
     }
 
-    public void testPartialResponseHandling() throws SQLException {
+    public void testPartialResponseHandling() throws Exception {
```
[the remark this answers is not in the document]
Comment (bpintea) [reply x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/JdbcShardFailureIT.java:65]: Just to make sure: this is just the testing, not the driver code. (The IOException that triggered exception promotion can be raised when setting up the test data.)

File: x-pack/plugin/sql/sql-action/src/main/java/org/elasticsearch/xpack/sql/action/SqlQueryRequest.java [hunk 5]
```diff
@@ -309,7 +330,8 @@ public boolean equals(Object obj) {
             && binaryCommunication == ((SqlQueryRequest) obj).binaryCommunication
             && Objects.equals(waitForCompletionTimeout, ((SqlQueryRequest) obj).waitForCompletionTimeout)
             && keepOnCompletion == ((SqlQueryRequest) obj).keepOnCompletion
-            && Objects.equals(keepAlive, ((SqlQueryRequest) obj).keepAlive);
+            && Objects.equals(keepAlive, ((SqlQueryRequest) obj).keepAlive)
+            && Objects.equals(allowPartialSearchResults, ((SqlQueryRequest) obj).allowPartialSearchResults);
```
[the remark this answers is not in the document]
Comment (bpintea) [reply x-pack/plugin/sql/sql-action/src/main/java/org/elasticsearch/xpack/sql/action/SqlQueryRequest.java]: Switched to equality and reshuffled comparisons.

File: x-pack/plugin/sql/sql-client/src/main/java/org/elasticsearch/xpack/sql/client/HttpClient.java [hunk 6]
```diff
@@ -173,7 +175,8 @@ private boolean head(String path, long timeoutInMs) throws SQLException {
             cfg.authUser(),
             cfg.authPass(),
             cfg.sslConfig(),
-            cfg.proxyConfig()
+            cfg.proxyConfig(),
+            cfg.allowPartialSearchResults()
```
[the remark this answers is not in the document]
Comment (bpintea) [reply x-pack/plugin/sql/sql-client/src/main/java/org/elasticsearch/xpack/sql/client/HttpClient.java]: Replaced with the default `false` value.

File: x-pack/plugin/sql/sql-proto/src/main/java/org/elasticsearch/xpack/sql/proto/SqlQueryRequest.java [hunk 7]
```diff
@@ -245,7 +257,8 @@ public boolean equals(Object o) {
             && Objects.equals(binaryCommunication, that.binaryCommunication)
             && Objects.equals(waitForCompletionTimeout, that.waitForCompletionTimeout)
             && keepOnCompletion == that.keepOnCompletion
-            && Objects.equals(keepAlive, that.keepAlive);
+            && Objects.equals(keepAlive, that.keepAlive)
+            && Objects.equals(allowPartialSearchResults, that.allowPartialSearchResults);
```
[the remark this answers is not in the document]
Comment (bpintea) [reply x-pack/plugin/sql/sql-proto/src/main/java/org/elasticsearch/xpack/sql/proto/SqlQueryRequest.java]: Addressed.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/CompositeAggCursor.java [hunk 8]
```diff
@@ -134,7 +134,7 @@ public void nextPage(SqlConfiguration cfg, Client client, ActionListener<Page> l
             log.trace("About to execute composite query {} on {}", StringUtils.toString(nextQuery), indices);
         }
 
-        SearchRequest request = prepareRequest(nextQuery, cfg.requestTimeout(), includeFrozen, indices);
+        SearchRequest request = prepareRequest(nextQuery, cfg.requestTimeout(), includeFrozen, cfg.allowPartialSearchResults(), indices);
```
[the remark this answers is not in the document]
Comment (bpintea) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/CompositeAggCursor.java]: `cfg` object passed as parameter now.

File: x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/JdbcShardFailureIT.java [hunk 9]
```diff
@@ -60,10 +62,79 @@ public void createTestIndex() throws IOException {
...
+
+        for (int i = 0; i < maxWarningHeaders - 1 + okShards + extraBadShards; i++) {
+            Request request = new Request("PUT", "/test" + i);
+            boolean indexWithDocVals = i < okShards;
+            request.setJsonEntity(String.format(Locale.ROOT, mappingTemplate, indexWithDocVals, indexWithDocVals));
+            assertOK(provisioningClient().performRequest(request));
+
+            request = new Request("POST", "/test" + i + "/_doc");
+            request.setJsonEntity("{\"bool\": " + (indexWithDocVals || randomBoolean()) + "}");
+            assertOK(provisioningClient().performRequest(request));
```
[the remark this answers is not in the document]
Comment (bpintea) [reply x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/JdbcShardFailureIT.java]: `indexName` extracted and reused.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/querydsl/container/QueryContainer.java [hunk 10]
```diff
@@ -116,7 +117,7 @@ public Attribute attribute() {
     private Map<Attribute, FieldAttribute> fieldAlias;
 
     public QueryContainer() {
-        this(null, null, null, null, null, null, null, -1, false, false, -1);
+        this(null, null, null, null, null, null, null, -1, false, false, -1, false);
```
[the remark this answers is not in the document]
Comment (bpintea) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/querydsl/container/QueryContainer.java]: Switched to `CoreProtocol` defaults, but only one more aditional value `INDEX_INCLUDE_FROZEN` could be reused from there.

**Comment by bpintea** [comment]
> Also, have you considered the CLI in any way for this new parameter, or it doesn't make sense?

I think it makes sense to keep the CLI in sync, but didn't want to extend this PR with it. I'll open an issue once this is merged.

**Pull request merged**